nixos/modules/security/sgx.nix

```diff
@@ -0,0 +1,167 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let
+  cfg = config.security.sgx;
+in
+{
+  options.security.sgx = {
+    enable = mkEnableOption "Intel SGX";
+
+    config = mkOption {
+      type = types.lines;
+      default = "";
+      description = "Configuration for the SGX daemon.";
+    };
+
+    packages = {
+      psw = mkOption {
+        default = pkgs.intel-sgx-psw;
+        defaultText = "pkgs.intel-sgx-psw";
+        description = "The SGX PSW (platform software) package to use.";
+      };
+
+      sdk = mkOption {
+        default = pkgs.intel-sgx-sdk;
+        defaultText = "pkgs.intel-sgx-sdk";
+        description = "The SGX SDK (software development kit) package to use.";
+      };
+
+      driver = mkOption {
+        type = types.package;
+        default = pkgs.linuxPackages.isgx;
+        defaultText = "pkgs.linuxPackages.isgx";
+        description = "The SGX driver package to use.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    boot.extraModulePackages = [ cfg.packages.driver ];
+    services.udev.packages = [ cfg.packages.psw ];
+
+    # aesmd depends on auditd
+    security.auditd.enable = true;
+
+    # write service config file
+    environment.etc."aesmd.conf".text = cfg.config;
+
+    # create service user and groups
+    users = {
+      users.aesmd = {
+        group = "aesmd";
+        isSystemUser = true;
+        extraGroups = [ "sgx_prv" ];
+        description = "Intel SGX Service Account";
+      };
+
+      groups = {
+        aesmd = { };
+        sgx_prv = { };
+      };
+    };
+
+    systemd.services.aesmd =
+      let
+        path = "${cfg.packages.psw}/aesm";
+      in
+      {
+        description = "Intel(R) Architectural Enclave Service Manager";
+        after = [ "syslog.target" "network.target" "auditd.service" ];
+        wantedBy = [ "multi-user.target" ];
+
+        # restart when config file changes
+        restartTriggers = [ config.environment.etc."aesmd.conf".source ];
+
+        environment = {
+          NAME = "aesm_service";
+          AESM_PATH = "/var/opt/aesmd";
+          LD_LIBRARY_PATH = "${cfg.packages.psw}/lib:${path}";
+        };
+
+        serviceConfig = {
+          ExecStart = "${path}/aesm_service --no-daemon";
+          ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+          ExecStartPre = [ "${pkgs.coreutils}/bin/mkdir -p /var/opt/aesmd/data /var/opt/aesmd/fwdir/data" ];
+
+          Restart = "on-failure";
+          RestartSec = "15s";
+
+          # environment
+          User = "aesmd";
+          Group = "aesmd";
+
+          WorkingDirectory = "/var/opt/aesmd";
+          RuntimeDirectory = "aesmd";
+          RuntimeDirectoryMode = "0755";
+          ReadWritePaths = [ "/var/opt/aesmd" ];
+
+          # Hardening
```

Comment:
From my own experience with running aesmd and systemd hardening option, the option chosen here looks fine.
It looks scary of the high number but to be honest most services should not do the majority of options that were disabled here. I also prefer remove hardening afterwards rather than breaking existing setups by adding more hardening.

Reply:
Upstream probably don't care as much about adding them because they might target some old redhat linux that does not have those hardening options yet.